src/shared/dissect-image.c

```diff
@@ -529,91 +147,74 @@ static void check_partition_flags(
-                local_deadline = start +
-                        DIV_ROUND_UP(n - start, retrigger_timeout) * retrigger_timeout;
+        struct blkpg_ioctl_arg ba = {
+                .op = BLKPG_ADD_PARTITION,
+                .data = &bp,
+                .datalen = sizeof(bp),
+        };
 
-                if (deadline != USEC_INFINITY && deadline <= local_deadline) {
-                        local_deadline = deadline;
-                        last_try = true;
-                } else
-                        last_try = false;
+        if (strlen(name) >= sizeof(bp.devname))
+                return -EINVAL;
 
-                r = device_wait_for_initialization(device, subsystem, local_deadline, ret);
-                if (r >= 0 && DEBUG_LOGGING) {
-                        const char *sn = NULL;
+        strcpy(bp.devname, name);
 
-                        (void) sd_device_get_sysname(device, &sn);
-                        log_device_debug(device,
-                                         "Successfully waited for device '%s' to initialize for %s.",
-                                         strna(sn),
-                                         FORMAT_TIMESPAN(usec_sub_unsigned(now(CLOCK_MONOTONIC), start), 0));
+        return RET_NERRNO(ioctl(fd, BLKPG, &ba));
+}
 
-                }
-                if (r != -ETIMEDOUT || last_try)
-                        return r;
+static int make_partition_devname(
+                const char *whole_devname,
+                int nr,
+                char **ret) {
 
-                if (DEBUG_LOGGING)
-                        log_device_debug(device,
-                                         "Device didn't initialize within %s, assuming lost event. Retriggering device.",
-                                         FORMAT_TIMESPAN(usec_sub_unsigned(now(CLOCK_MONOTONIC), start), 0));
+        bool need_p;
 
-                r = sd_device_trigger(device, SD_DEVICE_CHANGE);
-                if (r < 0)
-                        return r;
-        }
-}
-#endif
+        assert(whole_devname);
+        assert(nr > 0);
 
-#define DEVICE_TIMEOUT_USEC (45 * USEC_PER_SEC)
+        /* Given a whole block device node name (e.g. /dev/sda or /dev/loop7) generate a partition device
+         * name (e.g. /dev/sda7 or /dev/loop7p5). The rule the kernel uses is simple: if whole block device
+         * node name ends in a digit, then suffix a 'p', followed by the partition number. Otherwise, just
+         * suffix the partition number without any 'p'. */
 
-static void dissected_partition_done(DissectedPartition *p) {
-        assert(p);
+        if (isempty(whole_devname)) /* Make sure there *is* a last char */
+                return -EINVAL;
 
-        free(p->fstype);
-        free(p->node);
-        free(p->label);
-        free(p->decrypted_fstype);
-        free(p->decrypted_node);
-        free(p->mount_options);
+        need_p = strchr(DIGITS, whole_devname[strlen(whole_devname)-1]); /* Last char a digit? */
```

Comment:
`need_p = isdigit(whole_devname[strlen(whole_devname) - 1])`?

Reply:
isdigit() sucks. It's locale dependent (i.e. there's even isdigit_l). We really want only ASCII 0…9 here, hence we avoid the function.

(I mean let's hope that all locales define that the same way, but already the possibility that they could depart from ASCII 0…9 and add some other number chars to it, sounds like something we really should avoid).

`DIGITS` is defined by us, we know exactly what it is...

Reply:
The man page for the libc isxxx() functions says that these functions are locale dependent but it also says "isdigit() checks for a digit (0 through 9)". The "0 through 9" part made me believe that this is less locale dependent than what it probably is...